### ISSUE TITLE
HORRIBLE bug with namelist.echam runctl

### ISSUE THREAD

**Issue by dbarbi** [esm-tools/esm_tools#236]
**Describe the bug**
For some (also the default awicm ) settings, the first timestep in the output data is wrong. The problem is only really visible if you look at daily data, as the signal might be too small for monthly data. As far as we can see, potentially all echam standalone and awicm / awi-esm run potentially run into that problem; foci might be safe.

**To Reproduce**
Standard run with awicm-CMIP6 for example, with the standard namelist settings, look at the daily output.

**Screenshots**
look [here](https://spaces.awi.de/display/CSCPD/Echam6+rerun+issue) for more information

**System (please complete the following information):**
 - Supercomputer: doesn't matter
 - Version: all versions, probably going back to before ESM-Tools

**Additional context**
This is our understanding of what is going on as of Dec 4, 8:30 (all thanks to Karl-Hermann Wieners who is helping us a lot on this):

It appears as if there are two distinct problems, but very similar. The problem seems to be that a number of settings in the namelist.echam need to fit together perfectly, none of which is really documented in the echam manual by the way:

in runctl:
- putrerun: this setting says at which timestep restart files are written, important is the entry 'first' or 'last', saying whether the restart is written on the last timestep of the run, or formally on the first timestep of the next run
- putdata: same as putrerun, but for output data, not restarts
- dt_start: this is the date at which the event_manager of echam starts. Among other things, the event_manager controls the dates at which radiation is calculated. If missing, dt_start is taken from the restart_files to be the initial date (?) of the run. for short runs it is safe to omit this, but for extremely long runs, setting up the event_manager can then take a long time, so then it makes sense to set dt_start to a date in the near past of the current run.
- dt_resume: is the date at which the run is restarted; if missing, this is taken from the restart files if lresume is .True.
- default_output: specifies if the default output is written, and even though seemingly unrelated to the other settings, this guy is treacherous it seems (see below)

in mvstreamctl:
- interval: same as putdata, but for this stream.

so far, so good. now the following things can happen 

problem 1: 
if putrerun is set to 'last' and putdata/interval to 'first':
     restart is written at time 0,0,0 - timestep, putdata at 0,0,0 . Now the output data is a mean value, and it needs the last
     timestep from the restart file, which is missing. so the very last timestep in the output stream is messed up.
if putrerun is set to 'first' and putdata/interval to 'last':
    restart is written at 0,0,0, putdata at 0,0,0 - timestep. for the same reason as above, a timestep in the output streams is 
    corrupted, this time it is the first timestep.

potential problem 1a: 
if default_output is not found in the namelist.echam, or set to .True., it changes the default value for the 'interval' in mvstreamctl to 'first' !!! (WHY???)

BAD NEWS: All first time steps after a restart are affected.
maybe good news: The restart files should be Ok, and it might be possible to rescue the first timestep using a cdo script, or by redoing the december months

problem 2: 
dt_start is set to something that doesn't fit (as far as I know right now, that might be the case if everything is set to 'last', dt_resume is taken from the restart file maybe, and dt_start is set to time 0,0,0):
    the event_manager gets a hiccup, calculating strange timesteps for the radiation to be interpolated. that goes wrong, for 
    reasons I don't fully understand, but that leads to strange values in temperature and pressure level (that dissipate after the
    first timestep quickly)

BAD NEWS: similar looking as with problem 1, messed up values at the first timestep in some output fields

potential problem 3: 
dt_resume doesn't fit to 'first' / 'last' setting
    i actually don't know what happens then.

**Comment by dbarbi** [comment]
At the moment, a possible solution seems to be to make sure all these settings perfectly fit together:

putrerun, putdata and interval:
* if one setting is 'last', all of them should be 'last'
* if one setting is 'first', all of them should be 'first'

putrerun, dt_resume:
* if putrerun is 'first', make sure dt_resume is set to time 0,0,0
* if putrerun is 'last', make sure dt_resume is set to time 0,0,0 - timestep
(omit dt_resume? i don't know if that is safe)

dt_start:
* for short runs: omit this parameter
* if it can't be omitted: set dt_start to something in the near past of dt_resume

interval:
* make sure the 'interval' setting in mvstreamctl is always set explicitly, so that default_output can't interfere

**Comment by dbarbi** [comment]
Some good news:

All of the settings are already in foci.yaml. I suggest to move them to echam.yaml to make them the standard for everybody.
To take care of the 'interval' setting, I suggest a hotfix, to make a copy of each namelist.echam in the namelist folder, on with 'first', the other with 'last', and set in echam.yaml which one to take. not nice, but might help quickly.

**Comment by dbarbi** [comment]
Further explanation:

The 'last' setting is as a standard used by MPIESM and FOCI communities for climate research, to make sure that Nov 1, 0,0,0 is actually in the november mean and not the october one. Weather prediction uses 'first', because they want to restart exactly at 0,0,0. 
I don't know why, but the standard setting for awi-cm has always been 'first' for the reruns. That was always annoying if you wanted to re-use am echam restart file from a mpiesm experiment. 

AMIP runs actually have to use 'first', as that is how the forcing is prepared.

**Comment by dbarbi** [comment]
Additional information:

*dt_start and dt_resume should have the same time, so safest would be if dt_start is one year before dt_resume (if it has to be set at all).
* dt_resume ONLY NEEDS TO BE SET at the first year of a run, and only if you want to change the date that is in the restart file  you start from. For example if you have a spinup that goes until year 6000, and then want to reset that number to 1850 for pre-industrial runs. IN ALL OTHER CASES IT IS SAFE TO NOT SET dt_resume
* if dt_resume is set, the time needs to be identical with the one in the restart file!
* get_ocean and put_ocean ALSO have the setting 'first' and 'last'. DON'T TOUCH THOSE!!! (set to 'last' in all cases I know)

**Comment by dbarbi** [comment]
I have the namelists ready, they are in branch hotfix/horrible...

**Comment by dbarbi** [comment]
ok, i have a hotfix. when running awicm with restart_firstlast = 'last' i get for the first month:

```
&runctl
    dt_start = 1999, 12, 31, 23, 52, 30
    dt_stop = 2000, 2, 28, 23, 52, 30
    putrerun = 1, 'months', 'last', 0
    lfractional_mask = .false.
    lresume = .false.
    out_datapath = '/work/ollie/dbarbi/esm_yaml_tests//fl1/run_20000101-20000131/work/'
    out_expname = 'fl1'
    rerun_filetype = 4
    delta_time = 450
    putdata = 1, 'months', 'last', 0
    nproma = 8
    lcouple = .true.
    getocean = 1, 'days', 'last', 0
    putocean = 1, 'days', 'last', 0
    lcouple_co2 = .true.
    default_output = .false.
    dt_resume = 1999, 12, 31, 23, 52, 30
/
```
and with restart_firstlast = first:
```
&runctl
    dt_start = 2000, 1, 1, 0, 0, 0
    dt_stop = 2000, 2, 29, 0, 0, 0
    putrerun = 1, 'months', 'first', 0
    lfractional_mask = .false.
    lresume = .false.
    out_datapath = '/work/ollie/dbarbi/esm_yaml_tests//fl2/run_20000101-20000131/work/'
    out_expname = 'fl2'
    rerun_filetype = 4
    delta_time = 450
    putdata = 1, 'months', 'first', 0
    nproma = 8
    lcouple = .true.
    getocean = 1, 'days', 'last', 0
    putocean = 1, 'days', 'last', 0
    lcouple_co2 = .true.
    default_output = .false.
    dt_resume = 2000, 1, 1, 0, 0, 0
/

```
looks about right. I will now run a few test to see how the following months look.

**Comment by dbarbi** [comment]
there was still a little bug, now i get for the 'first':

month 1:
```
dt_start = 2000, 1, 1, 0, 0, 0
    dt_stop = 2000, 3, 1, 0, 0, 0
    putrerun = 1, 'months', 'first', 0
    lfractional_mask = .false.
    lresume = .false.
    out_datapath = '/work/ollie/dbarbi/esm_yaml_tests//fl8/run_20000101-20000131/work/'
    out_expname = 'fl8'
    rerun_filetype = 4
    delta_time = 450
    putdata = 1, 'months', 'first', 0
    nproma = 8
    lcouple = .true.
    getocean = 1, 'days', 'last', 0
    putocean = 1, 'days', 'last', 0
    lcouple_co2 = .true.
    default_output = .false.
    dt_resume = 2000, 1, 1, 0, 0, 0
```
month 2:
```
&runctl
    dt_start = 1999, 2, 1, 0, 0, 0
    dt_stop = 2000, 3, 1, 0, 0, 0
    putrerun = 1, 'months', 'first', 0
    lfractional_mask = .false.
    lresume = .true.
    out_datapath = '/work/ollie/dbarbi/esm_yaml_tests//fl8/run_20000201-20000229/work/'
    out_expname = 'fl8'
    rerun_filetype = 4
    delta_time = 450
    putdata = 1, 'months', 'first', 0
    nproma = 8
    lcouple = .true.
    getocean = 1, 'days', 'last', 0
    putocean = 1, 'days', 'last', 0
    lcouple_co2 = .true.
    default_output = .false.
/
```
As far as I can see that is what should be there, right?

**Comment by dbarbi** [comment]
ok there seems to be an issue related to jsbach here too...
Here's another little piece of the puzzle (that was a bit tricky to find): in namelist.jsbach , hydrology_ctl, there are two settings called gethd and puthd. they need to be set to:

'last':
gethd = 24, 'steps', 'exact', 0
puthd = 24, 'steps', 'exact', -450

'first':
remove both values from the namelist

**Comment by dbarbi** [comment]
also, it is not that easy to run awicm with putrerun = 'last'. With all settings correct, I now get an error message from oasis telling me that the runs exceeds the maxtime. don't know how but it seems oasis starts counting at echams start, and never reaches fesoms end, as time is over.
i am trying with a dirty hack on the namcouple at the moment, but will also start my weekend in a minute

**Comment by dbarbi** [comment]
ok, here's some detail i didn't post before - default_output always writes instantaneos data, no means. so it is to be expected that the data in the default output should not be affected by the first/last problem

also, s. rast could not exclude the possibility that mvstream averaging might or might not use default_output data for the averaging, and wasn't sure what happens if the default output isn't present at all

**Comment by JanStreffing** [comment]
Seems to have been solved in #130 Closing here.
